### PR TITLE
Mrtpipelines update

### PR DESCRIPTION
**Pull request by kaitj**
## Proposed changes

A number of quality changes (primarily to `mrtpipelines.smk`) to try to bring this up-to-date as of `snakebids==0.6.2`. Includes updates to:

* pybids_inputs
* Use of variables instead of grabbing from config for every rule
* Renamed rules to reflect similar naming of mrtrix functions
* Update mrtrix container to 3.0.3.
* Make use of `functools.partial` 

See the commit notes for full details.

## Types of changes

What types of changes does your code introduce? Put an `x` in the boxes that apply

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionalitiy)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other (if none of the other choices apply)

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you are unsure about any of the choices, don't hesitate to ask!

- [ ] Changes have been tested to ensure that fix is effective or that a feature works.
- [ ] Changes pass the unit tests
- [x] Code has been run through the `poe quality` task
- [x] I have included necessary documentation or comments (as necessary)
- [ ] Any dependent changes have been merged and published

## Notes
All PRs will undergo the unit testing before being reviewed. You may be requested to explain or make additional changes before the PR is accepted.

### CODE REVIEW

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:11]: Typo: `Paramaters` -> `Parameters`

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:24]: I don't think these need to be input functions (i.e. I think you could get rid of `lambda wildcards: `

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:9]: As far as I can tell `config["derivatives"] == False` by default. Maybe `join(config["out_dir"], derivatives, "mrtrix")` or something like that?

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:90]: I think this would be a bit cleaner if you were to get rid of the if statement and provide a sensible default for `shell` and `lmax` where they're defined on line 13.

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:97]: Is there any reason not to write this as three rules (one for each region), or make the `desc` a wildcard?

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:392]: Missing `,` here

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:433]: I think this would be more readable by defining a param: `nodes=",".join(str(num) for num in range(2, 73))`

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:509]: Nitpick: The `desc` here isn't very BIDSy because it has a `-` and `_`. Maybe `desc=from{node1}to{node2}`?

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:151]: It would save a bit of space to define a variable outside the rule with the `desc` as a wildcard, then expand over it for each output. Something like:

```
dwi2fod_out = bids(
    root=mrtrix_dir,
    datatype="response",
    model="csd",
    desc="{desc}",
    suffix="fod.mif",
    **config["subj_wildcards"],
)

rule dwi2fod:
    ...
    output:
        wm_fod=expand(dwi2fod_out, desc="wm", allow_missing=True),
    ....
```

Comment (tkkuehn) [reply dbsc/workflow/rules/mrtpipelines.smk:151]: This applies for a few of the following rules, too.

**Review by tkkuehn — CHANGES_REQUESTED** [review]
I haven't actually tried running this yet, but a few comments from looking over the rules...

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:11]: > I haven't actually tried running this yet, but a few comments from looking over the rules...

Haven't actually added this .smk to link with the target yet because there are still parts that aren't working in this workflow.

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:90]: I agree that this isn't the cleanest. That said, I would rather avoid providing a default within the workflow as Mrtrix itself has a mechanism in their code to provide a default that works with the input diffusion data (e.g. dealing with 2 shells vs 3 shells without explicitly defining the number of shells).

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:24]: Fair - this was copied from another workflow. I just realized I don't think `re` is imported here either. I guess 2 things to fix:

1. `lambda wildcards:`
2. `import re`

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [reply dbsc/workflow/rules/mrtpipelines.smk:90]: Gotcha. In that case, could do:
```
params:
    shells=f"-shells {shells}" if shells else ""
    lmax=f"-lmax {lmax}" if lmax else ""
```

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:97]: A consequence of how the original function was implemented in mrtrix (I think). No reason not to change it here though. Wildcard is probably best, assuming we can have it perform for each input tissue type (I think I have some old code that use to do something similar for hemispheres - will take a look).

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:433]: Agreed - will make the change this morning. This was a result of copying over the command from the shell scripts originally.

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:151]: Would that still work to define it as a wildcard if the different tissue types still need to be explicitly defined in the script (in order of `{input} {output} {input} {output} ...`)

Comment (tkkuehn) [reply dbsc/workflow/rules/mrtpipelines.smk:151]: Yeah, to be clear I would still define each of `wm_fod`, `gm_fod`, and `csf_fod` as inputs, I would just define them this way instead of repeating the (almost identical) `bids` definition every time

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:90]: Would that not pass "" to the command? I was running into that issue before and that ended up causing issues.

Comment (tkkuehn) [reply dbsc/workflow/rules/mrtpipelines.smk:90]: I would change the shell directive to:
`"dwi2response dhollander {input.dwi} {output.wm_rf} {output.gm_rf} {output.csf_rf} -nthreads {threads} -mask {input.mask} {params.shells} {params.lmax} "`,
which I would expect to resolve to:
`"dwi2response dhollander {input.dwi} {output.wm_rf} {output.gm_rf} {output.csf_rf} -nthreads {threads} -mask {input.mask}  "`
if `params.shells` and `params.lmax` were both an empty string, but it's possible I'm incorrect.

Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:90]: Well, we will find out when we test this.

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:147]: Great, I went ahead and updates to do something similar to this for most of the pre-processing rules. Not sure if its the best idea yet for the tractography rules.

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:93]: Scratch the second half of what I wrote before. I just made this similar to the `dwi2fod` example you have below. We could make it its own separate rule for each tissue type, but I think given what it is doing, I would prefer to keep them together even if its just calling the same command 3 different times in the shell script.

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:222]: This is another rule that could be split into two.

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:477]: I don't know this part of mrtrix very well but seems like this should have a `&&` between the first and second lines (or split into two rules)?

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (tkkuehn) [inline dbsc/workflow/rules/mrtpipelines.smk:515]: This is another one that should have `&&` or be two rules.

**Review by tkkuehn — CHANGES_REQUESTED** [review]
A couple of things that I'm pretty sure will break the workflow, and some stylistic suggestions that you can take or leave depending on your preference. We should also put together some test data (even if just empty files) and make sure a dry-run of this completes before merging.

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:222]: Yes agreed, although again, not too much benefit and ultimately this could be an optional rule down the line (may not want tensor measures).

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:95]: This was loosely based off of the workflow here: https://github.com/kaitj/mrtproc/blob/main/mrtproc/workflow/Snakefile

I'd probably hold off on making this change until we have a chance to test.

File: dbsc/workflow/rules/mrtpipelines.smk
Comment (kaitj) [reply dbsc/workflow/rules/mrtpipelines.smk:92]: I think maybe I'm not following. Maybe something to better to discuss over a meeting?

**Comment by kaitj** [comment]
I just went through and replaced any use of expand for the old outputs to make use of `functools.partial`

**Pull request merged**
